### ISSUE TITLE
quelpa-upgrade does not finish successfully

### ISSUE THREAD

**Issue by vyp** [vyp/evil-quick-scope#1]
Making version-control local to evil-quick-scope-autoloads.el while let-bound!
    Generating autoloads for evil-quick-scope.el...done
    Saving file /home/rws/.emacs.d/elpa/evil-quick-scope-20150817.2316/evil-quick-scope-autoloads.el...
    Wrote /home/rws/.emacs.d/elpa/evil-quick-scope-20150817.2316/evil-quick-scope-autoloads.el
    (No changes need to be saved)
    async-bytecomp-get-allowed-pkgs: Wrong type argument: arrayp, nil

After this, updating just stops.

Doing:

``` elisp
(setq async-bytecomp-allowed-packages nil)
```

bypasses this and allows quelpa-upgrade to finish. But it is not the solution because it requires users to have that in their configuration file, which they obviously should not need to.

**Comment by vyp** [comment]
@steckerhalter @wasamasa: Just to be clear, I'm not saying this is an issue with quelpa, I'm pretty sure this is an issue related to my *own* package. But the main reason I am pinging you guys is because I'm still very new to elisp and I figured asking the quelpa developers would be the best chance at a start.

I understand that the error message says I have a wrong type argument somewhere, but where and how do I start looking? I have an autoload cookie just before defining the minor mode, but I'm not exactly sure if I need the autoload cookie? Is that the problem? Is removing the cookie the fix?

**Comment by wasamasa** [comment]
Use `M-x toggle-debug-on-error` and reproduce a full backtrace please.

**Comment by steckerhalter** [comment]
@vyp I don't think it's a problem with you package. It is probably again this issue: https://github.com/jwiegley/emacs-async/issues/47

People who are using helm are told to do this:

```cl
(require 'helm-config)
```

and `helm-config` is requiring `async-bytecomp` which in turn [advices](https://www.gnu.org/software/emacs/manual/html_node/elisp/Advising-Functions.html) `package--compile` which is a core Emacs function from `package.el`. Now this advice from `async-bytecomp` breaks quelpa and the `(setq async-bytecomp-allowed-packages nil)` was added in `async-bytecomp` to work around it. So much for the history.

I'll have another look as to what is the exact reason it occurs. I was skimming it then and opened the issue back then.

**Comment by steckerhalter** [comment]
I know what the problem is now. `async-bytecomp` is trying to get the package dependencies via `package-archive-contents` which is nil in our case because we don't install our packages from MELPA (or whatever).

So it seems like `async-bytecomp` is trying to check if it should byte-compile a package and tries to get the dependencies of that package to whitelist them and this is what fails because it cannot get those dependencies if a package has been installed just locally with `package-install-file` (like we do with quelpa).

**Comment by vyp** [comment]
@wasamasa Ah... I was trying to reproduce this for *quite* some time to get the full backtrace, and I only now saw https://github.com/jwiegley/emacs-async/commit/0b60259ca8e7553c7029bb3471a1d29c3893200e, which seems to have fixed this issue. But https://github.com/jwiegley/emacs-async/pull/51 is still open, and there is no comment there either, meaning no email or notification of it being fixed! Hence why I thought this was still unfixed and why I kept trying to reproduce it without knowing it was already fixed.

Because the first quelpa-upgrade I did, evil-quick-scope was already up to date, and hence quelpa did not do anything with it (this was a mistake by me I didn't realise at the time). But it must have updated to the fixed version of async in this first upgrade, meaning all my next attempts at reproducing failed.

@steckerhalter Thanks very much for the help and looking into it further. Looks like you were correct.
